content/asa/0003-tecnicas-algoritmos.md

```diff
@@ -523,6 +522,169 @@ int activitySelection(std::vector<int> start, std::vector<int> finish) {
+
+:::
+
+:::details[Porque é que um código não pode ser prefixo de nenhum dos outros códigos?]
+
+Tomemos como exemplo um alfabeto de caracteres $\{a, b, c\}$. Poderíamos, teoricamente, criar um código binário tal que:
+
+$$
+a = 0\\
+b = 1\\
+c = 10
+$$
+
+Contudo, este código produz resultados **não determinísticos**, isto é, há mais que uma combinação de caracteres possível para uma dada sequência de bits, considerando os códigos em questão.  
+Consideremos, por exemplo, $010$ - podemos, aqui, ter $aba$ ou $ac$ como combinações de caracteres possíveis, já que ambas coincidem com a sequência de bits apresentada. Tal não aconteceria se cada código não pudesse ser prefixo de outro código.
+
+:::
+
+Podemos representar códigos livres de prefixo através de **árvores binárias** - é, aliás, nelas que o algoritmo de Huffman assenta.
+
+O conjunto de caracteres, com código binário tal que
+
+$$
+a = 01\\
+b = 00\\
+c = 1
+$$
+
+é livre de prefixo, já que nenhum código é prefixo de outro. Podemos representar o código com uma árvore binária (de realçar que aqui optamos por o 0 ser à esquerda e o 1 à direita, mas depende da implementação do algoritmo - no enunciado dos exercícios que eventualmente saírem vem especificado o sentido a escolher):
+
+![Árvore binária - código de Huffman](./assets/0003-arvore-huffman.png#dark=1)
+
+Se seguirmos o "caminho" da árvore binária, podemos observar o código a ser construído.
+
+:::info[Custo do Código Binário]
+
+Seja $C$ um código binário sobre um alfabeto $\Sigma$ e $f: \Sigma \to \N_0$ uma função de frequências (dos caracteres). Nesta situação, o custo do código é dado por:
+
+$$
+B(C) = \sum_{i \in \Sigma} f(i) \cdot C(i)
+$$
+
+Como o código tem uma árvore binária, $T$, associada, podemos também exprimir o custo desta maneira:
+
+$$
+B(T) = \sum_{i \in \Sigma} f(i) \cdot d_T(i)
+$$
+
+Aqui, $d_T$ corresponde à profundidade de $i$ na árvore binária.
+
+:::
+
+Temos, portanto, que o algoritmo de Huffman quererá **minimizar** o custo do código binário. O produto de um caracter muito frequente deve, num modelo ótimo, ser realizado com uma profundidade menor.
+
+O algoritmo em si passa por, inicialmente, considerar todos os caracteres como nós, com valor = frequência. Cria-se uma **min priority queue** (com prioridade = valor), e inserem-se os nós na mesma. De seguida, extraem-se os dois nós com valor menor e somam-se os respetivos valores, criando um novo nó com valor igual a essa soma. Este nó é inserido na queue, e assim sucessivamente $|\Sigma| - 1$ vezes, onde $|\Sigma|$ corresponde ao tamanho do alfabeto. O pseudocódigo é o seguinte:
+
+```
```

Comment:
```suggestion
```rust
```
Use `rust` as a language for this (pseudo-)code block since it's the one that gives the best syntax highlighting
![image](https://user-images.githubusercontent.com/7467891/146847212-b73d022d-50bb-42d3-bc23-6e0eae57a996.png)

Reply:
Will keep that in mind in the future, thanks for the tip!